### PR TITLE
[ODBC SCAN NODE] 4/4 Add ODBC_SCAN_NODE and Odbc_Scanner in BE and add ODBC_SCAN_NODE docs

### PR DESCRIPTION
**Pull request by HappenLee**
#4376  issue
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: be/src/exec/odbc_scan_node.cpp [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                return Status::InternalError(ss.str());
+            } else {
+                    RETURN_IF_ERROR(
+                            write_text_slot(static_cast<char*>(column_data.target_value_ptr), column_data.strlen_or_ind, slot_desc, state));
+            }
+            j++;
+        }
+
+        // Before we fix the problem utf8 encode sql query in SQLexecDirect
+        // we need to check some filter can not encode in asii code, like chinese
```
Comment (wutiangan) [inline be/src/exec/odbc_scan_node.cpp:229]: asii --> ascii

File: be/src/exec/odbc_scan_node.cpp [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    // create new tuple buffer for row_batch
+    int tuple_buffer_size = row_batch->capacity() * _tuple_desc->byte_size();
+    void* tuple_buffer = _tuple_pool->allocate(tuple_buffer_size);
+
+    if (NULL == tuple_buffer) {
+        return Status::InternalError("Allocate memory failed.");
+    }
+
+    _tuple = reinterpret_cast<Tuple*>(tuple_buffer);
+    // Indicates whether there are more rows to process. Set in _hbase_scanner.next().
```
Comment (wutiangan) [inline be/src/exec/odbc_scan_node.cpp:170]: _hbase_scanner?

File: be/src/exec/odbc_scanner.cpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    // Allocate a statement handle
+    ODBC_DISPOSE(_dbc, SQL_HANDLE_DBC, SQLAllocHandle(SQL_HANDLE_STMT, _dbc, &_stmt), "alloc statement");
+
+    ODBC_DISPOSE(_stmt, SQL_HANDLE_STMT, SQLExecDirect(_stmt, (SQLCHAR*)(query.c_str()), SQL_NTS), "exec direct");
+    // How many columns are there */
+    ODBC_DISPOSE(_stmt, SQL_HANDLE_STMT, SQLNumResultCols(_stmt, &_field_num), "count num colomn");
+
+    LOG(INFO) << "execute success:" << query <<  " column count:" << _field_num;
+
+    // check materialize num equeal _field_num
```
Comment (wutiangan) [inline be/src/exec/odbc_scanner.cpp:109]: equeal  -> equal

File: be/src/exec/odbc_scanner.cpp [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+Status ODBCScanner::error_status(const std::string& prefix, const std::string& error_msg) {
+    std::stringstream msg;
+    msg << prefix << " Err: " << error_msg;
+    LOG(WARNING) << msg.str();
+    return Status::InternalError(msg.str());
+}
+
+std::string ODBCScanner::handle_diagnostic_record(SQLHANDLE      hHandle,
```
Comment (wutiangan) [inline be/src/exec/odbc_scanner.cpp:196]: add some comments

File: docs/zh-CN/extending-doris/odbc-of-doris.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+|   DATE   |  DATE  |                                   |
+|  FLOAT   |  FLOAT  |                                   |
+|   TINYINT   | TINYINT |  |
+|   SMALLINT  | SMALLINT |  |
+|   INT  | INT |  |
+|   BIGINT  | BIGINT |  |
+|   DOUBLE  | DOUBLE |  |
+|   DATE  | DATE |  |
+|   DECIMAL  | DECIMAL |  |
+
```
Comment (wutiangan) [inline docs/zh-CN/extending-doris/odbc-of-doris.md:126]: hll data type?
datetime data type?

File: docs/zh-CN/extending-doris/odbc-of-doris.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    目前Doris只适配了MySQL和Oracle，关于其他的数据库的适配工作正在规划之中，原则上来说任何支持ODBC访问的数据库都能通过ODBC外表来访问。如果您有访问其他外表的需求，欢迎修改代码并贡献给Doris。
+
+3. 什么场合适合通过外表访问
+
+    通常在外表数据量较小，少于100W条时，可以通过外部表的方式访问。由于外表无法发挥Doris在存储引擎部分的能力和会带来额外的网络开销，所以建议根据实际对查询的访问时延要求来确定是否通过外部表访问还是将数据导入Doris之中。
+
+4. 通过Oracle访问出现乱码
+
+   尝试在BE启动脚本之中添加如下参数：`export NLS_LANG=AMERICAN_AMERICA.AL32UTF8`， 并重新启动所有BE
```
Comment (wutiangan) [inline docs/zh-CN/extending-doris/odbc-of-doris.md:160]: 这个参数如果直接放到be的启动脚本里会有啥影响？

File: be/src/exec/odbc_scanner.cpp [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    }
+    // We want ODBC 3 support
+    ODBC_DISPOSE(_env, SQL_HANDLE_ENV, SQLSetEnvAttr(_env, SQL_ATTR_ODBC_VERSION, (void *) SQL_OV_ODBC3, 0), "set env attr");
+    // Allocate a connection handle
+    ODBC_DISPOSE(_env, SQL_HANDLE_ENV, SQLAllocHandle(SQL_HANDLE_DBC, _env, &_dbc), "alloc dbc");
+    // Connect to the Database
+    ODBC_DISPOSE(_dbc, SQL_HANDLE_DBC, SQLDriverConnect(_dbc, NULL, (SQLCHAR*)_connect_string.c_str(), SQL_NTS,
+                           NULL, 0, NULL, SQL_DRIVER_COMPLETE_REQUIRED), "driver connect");
+
+    LOG(INFO) << "connect success:";
```
Comment (wutiangan) [inline be/src/exec/odbc_scanner.cpp:89]: add more detail informaton in log

File: docs/zh-CN/extending-doris/odbc-of-doris.md (hunk 6 above)
Comment (HappenLee) [reply docs/zh-CN/extending-doris/odbc-of-doris.md:160]: 不会有什么影响，但是原则上不使用oracle数据库的话是用不到这个环境变量的

File: docs/zh-CN/extending-doris/odbc-of-doris.md (hunk 5 above)
Comment (HappenLee) [reply docs/zh-CN/extending-doris/odbc-of-doris.md:126]: mysql now do not support hll, i will add datetime type in doc

File: be/src/exec/odbc_scanner.cpp (hunk 7 above)
Comment (HappenLee) [reply be/src/exec/odbc_scanner.cpp:89]: ok

File: be/src/exec/odbc_scan_node.h [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    virtual Status prepare(RuntimeState* state);
+
+    // Start MySQL scan using _mysql_scanner.
+    virtual Status open(RuntimeState* state);
+
+    // Fill the next row batch by calling next() on the _mysql_scanner,
+    // converting text data in MySQL cells to binary data.
+    virtual Status get_next(RuntimeState* state, RowBatch* row_batch, bool* eos);
+
+    // Close the _mysql_scanner, and report errors.
```
Comment (morningman) [inline be/src/exec/odbc_scan_node.h:51]: Modify the comment

File: be/src/exec/odbc_scanner.cpp [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+                                    auto err_msg = std::string("Errro in") + std::string(op); \
+                                    return Status::InternalError(err_msg.c_str()); \
+                                }  \
+                            } \
+
+static constexpr uint32_t SMALL_COLUMN_SIZE_BUFFER = 100;
+// Now we only treat HLL, CHAR, VARCHAR as big column
+static constexpr uint32_t BIG_COLUMN_SIZE_BUFFER = 65535;
+
+static std::u16string utf8_to_wstring (const std::string& str)
```
Comment (morningman) [inline be/src/exec/odbc_scanner.cpp:42]:
```suggestion
static std::u16string utf8_to_wstring(const std::string& str)
```

File: docs/zh-CN/extending-doris/odbc-of-doris.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+"table" = "baseall",
+"driver" = "Oracle 19 ODBC driver",
+"odbc_type" = "oracle"
+);
+```
+
+#### 2. 通过ODBC_Resource来创建ODBC外表 (推荐使用的方式)
+```
+create external resource "oracle_odbc"
+    properties
```
Comment (morningman) [inline docs/zh-CN/extending-doris/odbc-of-doris.md:72]: Indent

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**